core/src/main/scala/cr/pulsar/Topic.scala

```diff
@@ -50,17 +53,57 @@ object Topic {
     }
   }
 
-  def apply(cfg: Config, name: Topic.Name, typ: Type): Topic =
-    new Topic(
-      name,
-      URL(s"${typ.show}://${cfg.tenant.value}/${cfg.namespace.value}/${name.value}")
-    ) {}
+  private def buildUrl(cfg: Config, name: Name, `type`: Type): URL =
+    URL(s"${`type`.show}://${cfg.tenant.value}/${cfg.namespace.value}/${name.value}")
+
+  private case class TopicImpl(
+      name: Name,
+      url: URL,
+      cfg: Config
+  ) extends Topic {
+    def withType(_type: Type): Topic =
+      copy(url = buildUrl(cfg, name, _type))
+  }
+
+  /**
+    * It creates a topic with default configuration.
+    *
+    * - type: Persistent
+    */
+  def apply(name: String, cfg: Config): Topic = {
```

Comment:
Would it be better to expect `name: Name` here?

Reply:
I prefer the raw string for builders, seems more user friendly and having the newtype as an argument doesn't add much when there's no other similar type in the list of arguments

Reply:
Im imagining the usage like this:

`case class MyConfig(foo: Foo, topic: Topic.Name)` which I read using `ciris`, and then I create a consumer:
`Consumer(config.topic.value)`. 

Also, it could be confusing which part of the `Topic` we have to put here. Is it only `Topic.Name` or whole URL

Reply:
That's a good point :+1: